tests/integration/testvalidator.nim

```diff
@@ -0,0 +1,312 @@
+    check slotsNotFreed == slotsFailed  
+  
+  test "validator marks proofs as missing when using validation groups", NodeConfigs(
+    # Uncomment to start Hardhat automatically, typically so logs can be inspected locally
+    hardhat:
+      HardhatConfig.none,
+
+    clients:
+      CodexConfigs.init(nodes=1)
+        # .debug() # uncomment to enable console log output
+        .withLogFile() # uncomment to output log file to tests/integration/logs/<start_datetime> <suite_name>/<test_name>/<node_role>_<node_idx>.log
+        .withLogTopics("purchases", "onchain")
+        .some,
+
+    providers:
+      CodexConfigs.init(nodes=1)
+        .withSimulateProofFailures(idx=0, failEveryNProofs=1)
+        # .debug() # uncomment to enable console log output
+        # .withLogFile() # uncomment to output log file to tests/integration/logs/<start_datetime> <suite_name>/<test_name>/<node_role>_<node_idx>.log
+        # .withLogTopics("sales", "onchain")
+        .some,
+
+    validators:
+      CodexConfigs.init(nodes=2)
+        .withValidationGroups(groups = 2)
+        .withValidationGroupIndex(idx = 0, groupIndex = 0)
+        .withValidationGroupIndex(idx = 1, groupIndex = 1)
+        # .debug() # uncomment to enable console log output
+        .withLogFile() # uncomment to output log file to tests/integration/logs/<start_datetime> <suite_name>/<test_name>/<node_role>_<node_idx>.log
+        .withLogTopics("validator") # each topic as a separate string argument
+        .some
+  ):
+    let client0 = clients()[0].client
+    let expiry = 5.periods
+    let duration = expiry + 10.periods
+
+    # let mine a block to sync the blocktime with the current clock
+    discard await ethProvider.send("evm_mine")
+
+    var currentTime = await ethProvider.currentTime()
+    let requestEndTime = currentTime.truncate(uint64) + duration
+
+    let data = await RandomChunker.example(blocks=8)
+    
+    # TODO: better value for data.len below. This TODO is also present in
+    # testproofs.nim - we may want to address it or remove the comment.
+    createAvailabilities(data.len * 2, duration)
+
+    await marketplace.startTrackingEvents()
+
+    let cid = client0.upload(data).get
+    let purchaseId = await client0.requestStorage(
+      cid,
+      expiry=expiry,
+      duration=duration,
+      nodes=nodes,
+      tolerance=tolerance,
+      proofProbability=proofProbability
+    )
+    let requestId = client0.requestId(purchaseId).get
+
+    debug "validation suite", purchaseId = purchaseId.toHex, requestId = requestId
+
+    check eventuallyS(client0.purchaseStateIs(purchaseId, "started"),
+      timeout = (expiry + 60).int, step = 5)
+    
+    # if purchase state is not "started", it does not make sense to continue
+    without purchaseState =? client0.getPurchase(purchaseId).?state:
+      fail()
+      return
+    
+    debug "validation suite", purchaseState = purchaseState 
+
+    if purchaseState != "started":
+      fail()
+      return
```

Comment:
`check eventually` does not make the test fail. It just reports that check has failed and it goes further.

Reply:
Ah ok, you want to short cut the test execution in case of failure. Could you check the condition and return instead?
```nim
if not eventuallyS(client0.purchaseStateIs(purchaseId, "started"),
     timeout = (expiry + 60).int, step = 5):
   fail()
   return